### ISSUE TITLE
[css-ui?] Define interaction of block layout and editing.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#4904]
It seems all browsers except Gecko hack block layout so that an editable block has at least one non-empty line, see [this test-case](https://bug1625434.bmoattachments.org/attachment.cgi?id=9136291).

That is, other browsers render `<div contenteditable></div>` the same as `<div><br></div>`.

I'm on the fence on whether this is a browser bug on other browsers, or something that we should match and should be specified. If the later, how we do define this?

It seems other browsers only apply it to the editable root?

In blink this is done in [`LayoutBlock::HasLineIfEmpty`](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/layout/layout_block.cc;l=1656;drc=f1ddc02094084b931955fa2bf52cae1ebc2028f9), which seems inherited from WebKit's [`RenderBlock::hasLineIfEmpty`](https://webkit-search.igalia.com/webkit/rev/ab348309603e3a9f6609bba404bf498898bf1260/Source/WebCore/rendering/RenderBlock.cpp#2446).

I see the "do what I mean" intent of this, but it seems a bit of a layering violation / hack to me :(

cc @jfkthame @masayuki-nakano @kojiishi @rniwa

**Comment by emilio** [comment]
https://github.com/WebKit/webkit/commit/590ce138159ef25b1acfbfd1959f63aaa82b1d1a seems to be the first time this specialness was introduced.

**Comment by masayuki-nakano** [comment]
I think that `[contenteditable]:read-write` should have one line height at least by default UA stylesheet.

**Comment by emilio** [comment]
Maybe you can fix this in a non-hacky way with `min-height: 1em;`? But that only works if your box-sizing is `content-box` of course :/

**Comment by masayuki-nakano** [comment]
> Maybe you can fix this in a non-hacky way with `min-height: 1em;`? But that only works if your box-sizing is `content-box` of course :/

Well, but using `em` would cause changing the height when first character is typed.

And I wonder, there is same issue for empty inline element's width, isn't it? E.g., `<span contenteditable></span>`.

**Comment by jfkthame** [comment]
I'm also somewhat on the fence about this too; on the one hand, I think in an ideal world adding or removing contenteditable shouldn't cause any layout changes, but on the other hand that makes a simple `<div contenteditable></div>` essentially unusable.

So maybe something involving `min-height: 1lh` would be a reasonable hack.

**Comment by emilio** [comment]
So firefox has an [existing line-height hack](https://searchfox.org/mozilla-central/rev/b712398b7fae54ef377a558d6f16dede7a7f8530/layout/generic/nsLineLayout.cpp#2266) which applies when we find `white-space: pre` (or similar, see the `preMode` stuff in there).

This is pretty old code but it may be worth making it apply consistently for empty blocks, or something of that sort.

**Comment by emilio** [comment]
Ah, there's also https://github.com/w3c/editing/issues/70 for this.

**Comment by MatsPalmgren** [comment]
Fwiw, Chrome does it also for Grid containers for example:
`data:text/html,<div style="display:grid; border:solid" contenteditable></div>`

Also worth noting that a non-`auto` value has other side-effects for [Grid](https://drafts.csswg.org/css-grid/#min-size-auto)/[Flex](https://drafts.csswg.org/css-flexbox/#min-size-auto) items.

**Comment by emilio** [comment]
Yeah, chrome does it as a used-value-time hack. Also for flexbox: https://github.com/WebKit/webkit/commit/e293e4a4af3b41f2ad3d258cc9be17617ac165fd... :(

**Comment by rniwa** [comment]
As far as I know, forcing an empty contentediable to behave as if it had BR is a compatibility requirement.
